heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -0,0 +1,162 @@
+      throws ExecutionException, InterruptedException {
+    String topologyName = Runtime.topologyName(runtime);
+    PackingPlan existingPackingPlan = deserializer.fromProto(existingProtoPackingPlan);
+    PackingPlan proposedPackingPlan = deserializer.fromProto(proposedProtoPackingPlan);
+
+    Integer proposedContainerCount = proposedPackingPlan.getContainers().size();
+    Integer existingContainerCount = existingPackingPlan.getContainers().size();
+    Integer containerDelta = proposedContainerCount - existingContainerCount;
+
+    assertTrue(proposedContainerCount > 0,
+        "proposed packing plan must have at least 1 container %s", proposedPackingPlan);
+    assertTrue(containerDelta == 0 || scalableScheduler.isPresent(),
+        "Topology change requires scaling containers by %s but scheduler does not support "
+        + "scaling, aborting. Existing packing plan: %s, proposed packing plan: %s",
+        containerDelta, existingPackingPlan, proposedPackingPlan);
+
+    SchedulerStateManagerAdaptor stateManager = Runtime.schedulerStateManagerAdaptor(runtime);
+
+    // assert found PackingPlan is same as existing PackingPlan.
+    PackingPlans.PackingPlan foundPackingPlan = stateManager.getPackingPlan(topologyName);
+    assertTrue(foundPackingPlan.equals(existingProtoPackingPlan),
+        "Existing packing plan received does not equal the packing plan found in the state "
+        + "manager. Not updating updatedTopology. Received: %s, Found: %s",
+        existingProtoPackingPlan, foundPackingPlan);
+
+    // fetch the topology, which will need to be updated
+    TopologyAPI.Topology updatedTopology = stateManager.getTopology(topologyName);
+    Map<String, Integer> proposedComponentCounts = proposedPackingPlan.getComponentCounts();
+    for (String componentName : proposedComponentCounts.keySet()) {
+      Integer parallelism = proposedComponentCounts.get(componentName);
+      updatedTopology = mergeTopology(updatedTopology, componentName, parallelism);
+    }
+
+    // request new resources if necessary. Once containers are allocated we should make the changes
+    // to state manager quickly, otherwise the scheduler might penalize for thrashing on start-up
+    if (containerDelta > 0 && scalableScheduler.isPresent()) {
+      scalableScheduler.get().addContainers(containerDelta);
+    }
+
+    // update parallelism in updatedTopology since TMaster checks that
+    // Sum(parallelism) == Sum(instances)
+    logFine("Deleted existing Topology: %s", stateManager.deleteTopology(topologyName));
+    logFine("Set new Topology: %s", stateManager.setTopology(updatedTopology, topologyName));
+
+    // update packing plan to trigger the scaling event
+    logFine("Deleted existing packing plan: %s", stateManager.deletePackingPlan(topologyName));
+    logFine("Set new PackingPlan: %s",
+        stateManager.setPackingPlan(proposedProtoPackingPlan, topologyName));
+
```

Comment:
I agree that we should strive towards atomicity. The change described will we a considerable refactorting effecting much of the non-scaling functionality (i.e., distribution/update of physical plan, topology, etc during submit) so I'm inclined to ship an initial version of scaling that is weak transitionally but keeps submit as-is and then follow up with a refactoring to harden things from a transactional aspect.